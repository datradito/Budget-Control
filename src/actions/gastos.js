import uuid from 'uuid';
import database from '../firebase/firebase';


//EXPORTO POR NOMBRE, LUEGO ELIJO AL IMPORTAR
//ADD_GASTO - action generator
/* const funcAction =  ({} = {}) => ( {} ) */

export const addGasto =  (gasto) => ({
    type: 'ADD_GASTO',
    gasto
});

//ACCION COMPATIBLE PARA ENVIAR CON THUNK
export const startAddGasto = (( gastoDato = {} ) => {
    return (dispatch) => {
        const {
            descrip= '',
            note= '',
            importe= 0, 
            creadoAt= 0,
        } = gastoDato;

        const gasto = { descrip, note, importe, creadoAt };

        return database.ref('gastos').push(gasto)
        //pasando ref tengo acceso al .key desde firebase
        .then((ref) => {
            dispatch(addGasto({
                id: ref.key,
                ...gasto
            }));
        });
    };
});

//REMOVE_GASTO
export const removeGasto = ( { id } ) => ({
        type: 'REMOVE_GASTO',
        id
    
    });
//EDIT_GASTO
//no asigno default xq sino no tengo id/updates no hago nada en el reducer
export const editGasto = ( id, updates ) => ({
        type: 'EDIT_GASTO',
        id,
        updates
    }
);
    


//CODIGO PARA thunk
// ( gasto ) => ({
//         type: 'ADD_GASTO',
//         //gasto
//         //QUITO OBJETO XQ PASO TODO A FIREBASE
//         gasto: {
//             //id generado auto by uuid@3.1.0
//             id: uuid(),
//             //key:value iguales entonces simplifico
//             descrip,
//             note,
//             importe,
//             creadoAt
//         }
//     });



