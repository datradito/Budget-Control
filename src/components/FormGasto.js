import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


export default class FormGasto extends React.Component{
    //COMO EN OTRAS OCACIONES DEFINO SUPER PARA ACCEDER AL STATE
    constructor(props) {
        super(props);
        
        this.state = {
            //Si estoy editando, el state se llena con las props que se pasan a 
            //el componente <FormGasto/> si no, lleno el state con los valores por
            //defecto, ej: nota: props.gasto ? props.gasto.nota : '',
            descrip: props.gasto ? props.gasto.descrip : '',
            nota: props.gasto ? props.gasto.nota : '',
            importe: props.gasto ? (props.gasto.importe / 100).toString() : '',
            creadoAt: props.gasto ? moment(props.gasto.creadoAt) : moment(),
            calendarFocused: false,
            error:'',
        }
    }

    onChangeDescrip = (e) => {  
        const descrip = e.target.value;
        this.setState( () => ( { descrip } ) );
    };

    onChangeNota = (e) => {
        const nota = e.target.value;
        this.setState( () => ( { nota } ) );
    };

    onChangeImporte = (e) => {
        const importe = e.target.value;
        
        //REGULAR EXRPESSIONS, ver regex101.com
        if (!importe || importe.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState( () => ( { importe } ) );
        }
    };

    onDateChange = (creadoAt) => {
        if (creadoAt) {
            this.setState( () => ({ creadoAt }) );
    
        }
    };

    onFocusChange = ( {focused} ) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    };

    onSubmit = (e) => {
        //Prevengo refresh del explorador
        e.preventDefault();

        if(!this.state.importe || !this.state.descrip) {
            this.setState( () => (
                {error: 'Por favor ingrese descripcion e importe...'}
            ) )
        } else {
            this.setState( ()=>({ error:'' }));
            //props.onSubmit vienen desde el componente AddGastosPage pasada como props
            this.props.onSubmit({
                descrip: this.state.descrip,
                nota: this.state.nota,
                importe: parseFloat(this.state.importe, 10)*100, //CONVIERTO EL STRING importe EN NUMERO
                creadoAt: this.state.creadoAt.valueOf(), //USO FUNCION DE moment.js / timestamp
            })
        }
    };

    render () {
        return (
        <div>
            {/*Muestro error si es TRUE*/}
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                <input
                    type="text"
                    placeholder="Descripcion"
                    autoFocus
                    value={this.state.descrip}
                    onChange={this.onChangeDescrip}
                />
                <input
                    type="text"
                    placeholder="Importe"
                    value={this.state.importe}
                    onChange={this.onChangeImporte}
                />
                
                {/*https://github.com/airbnb/react-dates*/} 
                <SingleDatePicker 
                    //PROPS:
                    date= {this.state.creadoAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}

                    //SET PARA VER BOTON DE clear EN CALENDARIO, DEFAULT ES false
                    showClearDates = {() => {true}}
                    //SET PARA VER UN SOLO CALENDARIO LUEGO DE ELEGIR RANGO
                    numberOfMonths={1}
                    //SET PARA VER CALENDARIO DEL PASADO LUEGO DE ELEGIR RANGO
                    isOutsideRange={() => (false)}
                />
                
                <textarea
                    placeholder="Ingresar nota para el Gasto (opcional)"
                    value={this.state.nota}
                    onChange={this.onChangeNota}
                />
                <button>Agregar Gasto</button>
            </form>
        </div>
        )}
}