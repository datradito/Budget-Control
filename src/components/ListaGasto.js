import React from 'react';

//conecto el componente al store
import { connect } from 'react-redux';

import ListaItemGasto from './ListaItemGasto';

//importo el selector para renderear datos en orden, pongo el nombre q quiera
import selectedGastos from '../selectors/gastos';
/*Cuando mapeo el objeto gastos, este ya trae el estado del store tanto gastos como filters */

/** STATELESS FUNCTIONAL COMPONENT:
    *MAPEO EL ARRAY DE OBJ
    *Por cada objeto dentro de gastos, devuelvo una instancia del componente <ListaItemGasto />
    *paso el key, y con ...gasto envio todas las props (SPREAD) a
    *ListaItemGasto, de manera de no usar props.descripcion x ej
*/
export const ListaGasto = (props) => (
    <div>
        {
            props.gastos.length === 0 ? (
                <h2>Sin Gastos</h2>
            ) : (
                props.gastos.map((gasto) => {
                    return <ListaItemGasto key={gasto.id} {...gasto}/>
                })
            )
        }

    </div>
);

//MAPEO EL store state A props DEL COMPONENTE:
const mapStatetProp = (state) => {
    return {
        gastos: selectedGastos(state.gastos, state.filters)
    };
};

export default connect(mapStatetProp)(ListaGasto);