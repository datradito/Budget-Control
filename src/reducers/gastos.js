//Reducer Gastos

//Variables DEFAULT
//el default es un array, no un obj individual.
const gastosReducerDefault = [];

export default (state = gastosReducerDefault, action) => {
    switch (action.type) {
        case 'ADD_GASTO':
        //ES6 spread
        //[...state, {}] --->> el spread crea un nuevo array en memoria con el AGREGADO del objeto q envia dispatch(action), NO SE MODIFICA EL STATE!
            return [
                ...state,
                action.gasto
            ]
 
        case 'REMOVE_GASTO':
            //DEVUELVO SOLO LOS ELEMENTOS DISTINTOS AL id pasado a removeGasto 
            /* 
                id!==id -->> son distintos los idÂ¿?
                si true (son distintos) -->> sigo con el sgte item del array, no modifico nada
                si false (son iguales) -->> filtro el item, no lo incorporo al nuevo array.
            */
            return state.filter(({ id }) => id!== action.id);
            //IDEM
            /*return state.filter(({ id }) => {
                return id !== action.id
            }); */
        
        case 'EDIT_GASTO':
            //mapeo todas las props del obj gastos
            return state.map((gasto) => {
                //si encuentro el id a modificar
                if (gasto.id === action.id) {
                    //Entonces hago OBJ SPREAD OPERATOR para devolver el obj modificado
                    return {
                        ...gasto,
                        ...action.updates
                    };
                } else {
                    return gasto;
                };
            });

        default:
            return state;
    }
};