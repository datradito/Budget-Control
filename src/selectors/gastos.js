import moment from 'moment';

//Envio Gastos filtrados a la consola
//gastos(obj gastos, obj filters)
//obj filters : {text, sortBy, fechaInicio, fechaFin} 

export default (gastos, { text, sortBy, fechaInicio, fechaFin}) => {
    //.filter() devuelve un array con los match q dieron true para cada filtro:fechaInicioMatch, fechaFinMatch, textMatch.
    return gastos.filter((gasto) => {
        //Logica para filtrar dentro de fechaInicio-fechaFin
        const creadoAtMoment = moment(gasto.creadoAt);
        const fechaInicioMatch = fechaInicio ? fechaInicio.isSameOrBefore(creadoAtMoment, 'day') : true;            
        const fechaFinMatch = fechaFin ? fechaFin.isSameOrAfter(creadoAtMoment, 'day') : true;
        //filtro la descripcion si text esta incluida en la cadena , ver .includes()
        const textMatch = gasto.descrip.toLowerCase().includes(text.toLowerCase());

        return fechaInicioMatch && fechaFinMatch && textMatch;
        //ordeno por fecha de menor a mayor, e importe de mayor a menor. Ver  array.sort()
    }).sort((a, b ) => {
        //ordeno fecha
        if (sortBy === 'fecha') {
            return a.creadoAt < b.creadoAt ? 1 : -1; //ver compare en .sort()
        }
        //ordeno importe
        else if(sortBy === 'importe'){
             return a.importe < b.importe ? 1 : -1;
        }
    });
};

