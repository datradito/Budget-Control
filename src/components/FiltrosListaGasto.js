import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByImporte, sortByFecha, setFechaFin, setFechaInicio } from '../actions/filters';

/**
 * https://github.com/airbnb/react-dates
 */

export class FiltrosListaGasto extends React.Component{
    state = {
        //Necesito seguir el estado de 'calendarFocused' como en <FormGasto/>, default es null y luego pasa a string(calendario 1 o 2)
        calendarFocused: null
    };

    //SEGUN react-dates library
    onDatesChange = ({ fechaInicio , fechaFin}) => {
        this.props.setFechaInicio(fechaInicio);
        this.props.setFechaFin(fechaFin);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ( {calendarFocused} ))
    };

    onSortChange = (e) => {
        if(e.target.value==='date'){
            this.props.sortByFecha();
        }
        else if(e.target.value==='amount'){
            this.props.sortByImporte();
        }
    };

    onTextChange =  (e) => {
        this.props.setTextFilter(e.target.value);
    };

    render() {
        return (
            <div>
            <input 
                type="text" 
                //Tomo el valor text del filters en el store
                value={this.props.filters.text} 
                //Accedo a .dispatch() desde las props porque me lo envia connect()
                onChange={this.onTextChange} />

            <select 
                value={this.props.filters.sortBy}
                onChange={ this.onSortChange }
            >
                <option value="date">Fecha</option>
                <option value="amount">Importe</option>
            </select>

            <DateRangePicker
                startDate={this.props.filters.fechaInicio} // momentPropTypes.momentObj or null,
                endDate={this.props.filters.fechaFin} // momentPropTypes.momentObj or null,
                onDatesChange = {this.onDatesChange}
                //Toma el valor del state
                focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange= {() => false }
            />
        </div>
        );
    }
};

const mapStateToProps = (state) => ({
    filters : state.filters
});

const mapDispatchToProps = () => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByFecha: () => dispatch(sortByFecha()),
    sortByImporte: () => dispatch(sortByImporte()),
    setFechaInicio: (fechaInicio) => dispatch(setFechaInicio(fechaInicio)),
    setFechaFin: (fechaFin) => dispatch(setFechaFin(fechaFin)),
});


export default connect( mapStateToProps, mapDispatchToProps)( FiltrosListaGasto ) ;