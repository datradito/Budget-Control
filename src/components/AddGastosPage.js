import React from 'react';
import { connect } from 'react-redux';
import FormGasto from './FormGasto';
import { startAddGasto } from '../actions/gastos';

export class AddGastosPage extends React.Component {
    onSubmit = (gasto) => {
        this.props.startAddGasto(gasto);
        this.props.history.push('./'); 
    };
    render() {
        return (
            <div>
            <h1>Agrega Gasto</h1>
            <FormGasto onSubmit = {this.onSubmit}/>
        </div>
        );
    }
}

//Modifico el export para poder usar spy en Jest
/**
 * https://react-redux.js.org/using-react-redux/connect-mapdispatch
 */
const mapDispatchToProps = (dispatch) => ({
    startAddGasto: (gasto) => dispatch(startAddGasto(gasto))
})
export default connect(undefined, mapDispatchToProps)(AddGastosPage);



/**
 * ********************************************************************************
 * Documento el stateless component que se utilizaba antes de configurar para tests
 * ********************************************************************************
 */

//Al utilizar connect("state")("Componente") tengo acceso a .dispatch() dsd props. Ej: 
//const Componente = (props) => (jsx)

// const AddGastosPage = (props) => (
//     <div>
//         <h1>Agrega Gasto</h1>
//        <FormGasto 
//             //La prop OnSubmit = {} recibe dsd el componente <FormGasto /> el objeto con los datos para enviar al store, 
//             onSubmit = {(gasto) => {
//                 console.log(gasto);
//                 //props.dispatch(addGasto(gasto));
//                 props.onSubmit(gasto);
//                 //El componente recibe history kmo prop a traves de react-router, ver API.
//                 //con .push('/') redirijo al dashboard page.
//                 props.history.push('./'); 
//            }}
//        />
//     </div>

// );