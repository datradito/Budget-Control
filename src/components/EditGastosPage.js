import React from 'react';
import FormGasto from './FormGasto';
import { connect } from 'react-redux';
import { editGasto, removeGasto } from '../actions/gastos';

//ACCEDO A props QUE PASA EL COMPONENTE Route 
const EditGastosPage = (props) => {
    return (
        <div>
            <FormGasto 
                gasto= {props.gasto}
                onSubmit={ (gasto) => {
                    //dispatch(action())
                    props.dispatch(editGasto(props.gasto.id, gasto));
                    //redirijo a home
                    props.history.push('/');
                    console.log('Gasto actualizado', gasto);
                }}
            />
            <button 
                    onClick = { () => {
                    props.dispatch(removeGasto({id : props.gasto.id})); 
                    //redirect:
                    props.history.push('/');}}>
                    Borrar
            </button>
        </div>
    );
};


const mapsStateToProps = (state, props) => {
    return {
        //USO props.match.params PARA ACCEDER A LOS PARAMETROS DEL COMPONENTE <Route/>
        //find() javascript al array state para obtener el obj gasto con el id buscado
        gasto: state.gastos.find((gasto) => gasto.id === props.match.params.id)
    };
};

export default connect(mapsStateToProps)(EditGastosPage);