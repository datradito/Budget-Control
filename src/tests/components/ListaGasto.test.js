import React from 'react';
import { shallow } from 'enzyme';

//Importo ListaGasto por nombre, no el default export del componente
import { ListaGasto } from '../../components/ListaGasto';
import gastos from '../fixtures/gastos';

test('Deberia render ListaGasto con gastos', () => {
    const wrapper = shallow(<ListaGasto gastos={ gastos }/>);
    expect(wrapper).toMatchSnapshot();
})

test('Deberia render ListaGasto sin gastos', () => {
    const wrapper = shallow(<ListaGasto gastos={ [] }/>);
    expect(wrapper).toMatchSnapshot();
})