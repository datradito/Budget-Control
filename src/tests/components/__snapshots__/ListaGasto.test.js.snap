// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deberia render ListaGasto con gastos 1`] = `
<div>
  <ListaItemGasto
    creadoAt={0}
    descrip="Gas"
    id="1"
    importe={195}
    key="1"
    note=""
  />
  <ListaItemGasto
    creadoAt={-345600000}
    descrip="Luz"
    id="2"
    importe={109500}
    key="2"
    note=""
  />
  <ListaItemGasto
    creadoAt={345600000}
    descrip="Varios"
    id="3"
    importe={4500}
    key="3"
    note=""
  />
  <p>
    Cantidad de gastos: 
    3
  </p>
</div>
`;

exports[`Deberia render ListaGasto sin gastos 1`] = `
<div>
  <p>
    Sin Gastos
  </p>
  <p>
    Cantidad de gastos: 
    0
  </p>
</div>
`;
